# JBI100 Visualization 
### Academic year 2024-2025

## Incidents and Accidents
Data sources:

- Work-related Injury and Illness (https://www.osha.gov/Establishment-Specific-Injury-and-Illness-Data)


In [1]:
# Import libraries
import pandas as pd
import plotly.express as px
import numpy as np
import os

# Do not truncate tables
pd.set_option("display.max_columns", None)

# Assignment 1

## Exercise 1 – Data Set 

### (a) What is the information you can obtain from the data set/ data sets?

The OSHA Injury Tracking Application (ITA) Case Detail dataset contains detailed information on work-related injuries and illnesses (each row is a case of work-related injuries or illnesses). The data includes the following types of information:

#### 1. Establishment-Level Information
- **Unique Identifiers**: Establishment ID, Employer Identification Number (EIN).
- **Demographic Information**: Establishment name, company name, street address, city, state, zip code.
- **Industry Classification**: North American Industry Classification System (NAICS) code, year of NAICS code used, industry description.
- **Establishment Type**: Private industry, state government entity, or local government entity.
- **Workforce Data**: Size of the establishment, annual average employees, total hours worked.

#### 2. Incident-Level Information
- **Incident Identifiers**: Unique case number, establishment ID linkable to 300A data.
- **Incident Details**: Date of incident, type of incident (injury, skin disorder, etc.), time of incident, time started work prior to incident, and whether time was unknown.
- **Outcomes**: Most serious outcome (e.g., death, days away from work, job transfer/restriction), number of days away from work, number of restricted duty or transfer days.
- **Fatalities**: Date of death (if applicable).

#### 3. Narrative Descriptions
- **Incident Details**: What the employee was doing before the incident, how the incident happened, injury/illness description, and the object/substance directly harming the employee.

#### 4. Occupational Coding Information
- **Job Information**: Job title of the injured/ill employee.
- **Standard Occupation Code (SOC)**:
  - SOC Code and Description: Assigned using NIOCCS.
  - SOC Probability: Confidence score for SOC coding.
  - SOC Reviewed: Indicates whether the SOC code was reviewed or reassigned.

#### 5. System Metadata
- **Submission Information**: Created timestamp, year of filing.
- **Data Quality Indicators**: Codes for missing or invalid entries (e.g., "9999" for SOC code when unassignable).


### (b) What are the attributes in the data and what is their meaning?

The OSHA Injury Tracking Application (ITA) Case Detail dataset includes the following attributes, categorized by their context and meaning:

#### 1. **Establishment Information**
- **`establishment_ID`**: Unique identifier for each establishment.
- **`establishment_name`**: Name of the establishment reporting the data.
- **`ein`**: Employer Identification Number (Federal Tax Identification Number).
- **`company_name`**: Name of the parent company of the establishment.
- **`street_address`**: Street address of the establishment.
- **`city`**: City where the establishment is located.
- **`state`**: State or territory where the establishment is located.
- **`zip_code`**: Full zip code of the establishment.
- **`naics_code`**: North American Industry Classification System (NAICS) code for the establishment.
- **`naics_year`**: Year version of NAICS code used.
- **`industry_description`**: Industry description based on the NAICS code.
- **`establishment_type`**: Type of establishment:
  - 1 = Private industry
  - 2 = State government entity
  - 3 = Local government entity
- **`size`**: Size of the establishment based on maximum employees:
  - 1 = <20 employees
  - 21 = 20-99 employees
  - 22 = 100-249 employees
  - 3 = 250+ employees
- **`annual_average_employees`**: Annual average number of employees.
- **`total_hours_worked`**: Total hours worked by all employees at the establishment.

#### 2. **Incident Information**
- **`case_number`**: Employer-assigned unique case number for each injury/illness.
- **`date_of_incident`**: Date when the incident occurred.
- **`incident_outcome`**: Most serious outcome of the incident:
  - 1 = Death
  - 2 = Days away from work (DAFW)
  - 3 = Job transfer or restriction
  - 4 = Other recordable case
- **`dawf_num_away`**: Number of days away from work due to the incident.
- **`djtr_num_tr`**: Number of days on restricted duty or job transfer due to the incident.
- **`type_of_incident`**: Type of incident:
  - 1 = Injury
  - 2 = Skin disorder
  - 3 = Respiratory condition
  - 4 = Poisoning
  - 5 = Hearing loss
  - 6 = All other illness
- **`time_started_work`**: Time the employee began work prior to the incident.
- **`time_of_incident`**: Time the incident occurred.
- **`time_unknown`**: Indicator if the time of the incident is unknown:
  - 0 = No
  - 1 = Yes
- **`date_of_death`**: Date of death, if applicable.

#### 3. **Narrative Descriptions**
- **`incident_description`**: Description of the incident.
- **`nar_before_incident`**: Description of what the employee was doing before the incident.
- **`nar_what_happened`**: Description of what happened during the incident.
- **`nar_injury_illness`**: Description of the injury or illness.
- **`nar_object_substance`**: Description of the object or substance directly harming the employee.

#### 4. **Occupational Information**
- **`job_description`**: Job title of the injured/ill employee.
- **`SOC_code`**: Standard Occupation Code assigned by NIOCCS or OSHA.
- **`SOC_description`**: Text description of the SOC code.
- **`SOC_probability`**: Confidence score for the SOC coding (5 indicates a manually reassigned code).
- **`SOC_reviewed`**: Indicator of whether the SOC code was reviewed:
  - 0 = Not reviewed, NIOCCS coded
  - 1 = Reviewed by OSHA
  - 2 = Not SOC coded (SOC = "9999")

#### 5. **System Metadata**
- **`created_timestamp`**: Timestamp when the record was submitted.
- **`year_of_filing`**: Year in which the reported injuries/illnesses occurred.

### (c) Write a small parsing function that can read the data position (column, row) from the file format you selected. 

In [2]:
def parse_dataset():
    dataset_path = os.path.join(
        "Work-related Injury and Illness",
        "ITA Case Detail Data 2023 through 8-31-2023.csv",
    )
    return pd.read_csv(
        dataset_path,
        delimiter=",",
        low_memory=False,
        encoding="utf-8",
        dtype={"zip_code": "string", "naics_code": "string", "naics_year": "string"},
        index_col="id",
    )


df = parse_dataset()

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889447 entries, 446827 to 900939
Data columns (total 32 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   establishment_id          889447 non-null  int64  
 1   establishment_name        889447 non-null  object 
 2   ein                       812922 non-null  object 
 3   company_name              840391 non-null  object 
 4   street_address            889446 non-null  object 
 5   city                      889447 non-null  object 
 6   state                     889447 non-null  object 
 7   zip_code                  889447 non-null  string 
 8   naics_code                889447 non-null  string 
 9   naics_year                888174 non-null  string 
 10  industry_description      829648 non-null  object 
 11  establishment_type        887563 non-null  float64
 12  size                      889447 non-null  int64  
 13  annual_average_employees  889447 non-null  i

In [4]:
df.shape

(889447, 32)

In [5]:
df.sample(5)

establishment_id                      establishment_name        ein  \
id                                                                            
830838           1219338                                  WM 834  710862119   
798682            110200                                     XSD  942904084   
664759           1196690                                  Murray  391629977   
334843            167724    WellStar West Georgia Medical Center  581649541   
113270            925652  Northland Constructors of Duluth, Inc.  411451666   

                                  company_name         street_address  \
id                                                                      
830838                Wal-Mart Stores East, LP          810 S 37TH ST   
798682              XPO Logistics Freight Inc.         50 EDGEBORO RD   
664759                  Saputo Cheese USA, Inc  100 East Chestnut St.   
334843            WellStar Health System, Inc.       1514 Vernon Raod   
113270  Northland Constructors of Duluth, Inc.    4843 Rice Lake Road   

                  city state zip_code naics_code naics_year  \
id                                                            
830838         BETHANY    MO    64424     452910       2012   
798682  EAST BRUNSWICK    NJ    08816     484122       2022   
664759          Murray    KY    42071     311511       2022   
334843        LaGrange    GA    30240     622110       2022   
113270          Duluth    MN    55803     237310       2022   

                                     industry_description  establishment_type  \
id                                                                              
830838                   Warehouse Clubs and Supercenters                 1.0   
798682                             Freight Transportation                 1.0   
664759                 Flavored milk drinks manufacturing                 1.0   
334843            Hospitals, general medical and surgical                 1.0   
113270  Pavement, highway, road, street, bridge or air...                 1.0   

        size  annual_average_employees  total_hours_worked case_number  \
id                                                                       
830838    22                       156              276605    C3543737   
798682    22                       133              287126  4.29316E+1   
664759     3                       343              680639  INJ-000815   
334843     3                      2031             2411341      346175   
113270     2                       156              357897           8   

                    job_description soc_code  \
id                                             
830838  STOCKER GENERAL MERCHANDISE  53-7065   
798682             Driver Sales Rep  41-3091   
664759        Utility  Murray Union     9999   
334843                      ED Tech     9999   
113270              Cement Finisher  47-2071   

                                          soc_description  soc_reviewed  \
id                                                                        
830838                         Stockers and Order Fillers             1   
798682  Sales Representatives of Services, Except Adve...             1   
664759                                            Uncoded             2   
334843                                            Uncoded             2   
113270  Paving, Surfacing, and Tamping Equipment Opera...             0   

        soc_probability date_of_incident  incident_outcome  dafw_num_away  \
id                                                                          
830838         0.997296       01/08/2023                 4              0   
798682         0.943694       03/14/2023                 2            175   
664759         0.000000       07/13/2023                 2             49   
334843         0.000000       05/17/2023                 2              1   
113270         0.998629       09/28/2023                 4              0   

        dj

### (d) Write another function that outputs the distribution of the attributes, and counts the frequencies of the different values

In [6]:
df.describe(include="all")

establishment_id                    establishment_name        ein  \
count       8.894470e+05                                889447     812922   
unique               NaN                                 89219      33514   
top                  NaN  Walt Disney Parks and Resorts US Inc  820544687   
freq                 NaN                                  3532      38132   
mean        8.631224e+05                                   NaN        NaN   
std         3.575762e+05                                   NaN        NaN   
min         4.194000e+04                                   NaN        NaN   
25%         6.515120e+05                                   NaN        NaN   
50%         9.893690e+05                                   NaN        NaN   
75%         1.140101e+06                                   NaN        NaN   
max         1.246187e+06                                   NaN        NaN   

                   company_name            street_address     city   state  \
count                    840391                    889446   889447  889447   
unique                    38301                     88405    14317      56   
top     Amazon.com Services LLC  UNITED AIRLINES ATTN:MGR  Chicago      CA   
freq                      38132                      3748     5526  113262   
mean                        NaN                       NaN      NaN     NaN   
std                         NaN                       NaN      NaN     NaN   
min                         NaN                       NaN      NaN     NaN   
25%                         NaN                       NaN      NaN     NaN   
50%                         NaN                       NaN      NaN     NaN   
75%                         NaN                       NaN      NaN     NaN   
max                         NaN                       NaN      NaN     NaN   

       zip_code naics_code naics_year              industry_description  \
count    889447     889447     888174                            829648   
unique    17101        865          3                              5673   
top       32830     622110       2022  Warehouse Clubs and Supercenters   
freq       3586     155665     789067                             64327   
mean        NaN        NaN        NaN                               NaN   
std         NaN        NaN        NaN                               NaN   
min         NaN        NaN        NaN                               NaN   
25%         NaN        NaN        NaN                               NaN   
50%         NaN        NaN        NaN                               NaN   
75%         NaN        NaN        NaN                               NaN   
max         NaN        NaN        NaN                               NaN   

        establishment_type           size  annual_average_employees  \
count        887563.000000  889447.000000              8.894470e+05   
unique                 NaN            NaN                       NaN   
top                    NaN            NaN                       NaN   
freq                   NaN            NaN                       NaN   
mean              1.083898       9.833503              8.200306e+03   
std               0.362637       9.040229              1.033831e+06   
min               0.000000       1.000000              0.000000e+00   
25%               1.000000       3.000000              1.490000e+02   
50%               1.000000       3.000000              3.470000e+02   
75%               1.000000      22.000000              1.323000e+03   
max               3.000000      22.000000              1.723076e+08   

        total_hours_worked case_number        job_description soc_code  \
count         8.894470e+05      889434                 885793   889447   
unique                 NaN      423945                  93398      669   
top                    NaN           1  Fulfillment Associate     9999   
freq                   NaN       28102                  21562   157739   
mean          3.528374

In [7]:
def get_frequency_distribution(df):
    """
    Creates a DataFrame where:
    - Primary index: column names (attributes)
    - Secondary index: unique values in each column
    - Value column: frequency of each unique value

    Parameters:
    - df (pd.DataFrame): Input DataFrame.

    Returns:
    - pd.DataFrame: Frequency distribution as described.
    """
    frequency_df = pd.concat(
        {col: df[col].value_counts() for col in df.columns},
        names=["Attribute", "Value"],
    ).reset_index(name="Frequency")

    # Set the index as required
    return frequency_df.set_index(["Attribute", "Value"])


df_frequency = get_frequency_distribution(df)
df_frequency

Frequency
Attribute         Value                      
establishment_id  225446                 3532
                  857258                 2948
                  899222                 2218
                  62625                  1412
                  1126488                1267
...                                       ...
created_timestamp 22JAN24:16:48:00          1
                  27FEB24:11:49:00          1
                  13FEB24:01:22:00          1
                  30AUG24:17:57:00          1
year_filing_for   2023                 889447

[1108932 rows x 1 columns]

In [8]:
def attribute_distribution(dataframe, attribute, plot_distribution=False):
    """
    Calculates the distribution of values for a specified attribute in the dataset
    and optionally plots the distribution using Plotly Express.

    Parameters:
        dataframe (pd.DataFrame): The DataFrame containing the dataset.
        attribute (str): Column name of the attribute to analyze.
        plot_distribution (bool): Whether to plot the distribution of the attribute.

    Returns:
        pd.DataFrame: A DataFrame with value counts and percentage distribution.
    """
    if attribute not in dataframe.columns:
        raise ValueError(f"Attribute '{attribute}' not found in the dataset.")

    # Calculate value counts and percentage
    counts = dataframe[attribute].value_counts()
    percentages = (counts / counts.sum()) * 100

    # Combine counts and percentages into a DataFrame
    distribution = pd.DataFrame(
        {
            "Value": counts.index.astype(
                str
            ),  # Ensure all values are strings for categorical plotting
            "Frequency": counts.values,
            "Percentage": percentages.values,
        }
    ).set_index("Value")

    if not plot_distribution:
        return distribution

    fig = px.bar(
        distribution.reset_index(),
        x="Value",
        y="Percentage",
        text="Percentage",
        title=f"Distribution of {attribute} (Percentage)",
        labels={"Value": "Attribute Value", "Percentage": "Percentage (%)"},
    )
    fig.update_traces(texttemplate="%{text:.2f}%", textposition="outside")
    fig.update_layout(
        xaxis=dict(title="Values"),
        yaxis=dict(title="Percentage (%)"),
        uniformtext_minsize=8,
        uniformtext_mode="hide",
    )
    fig.show()

    return distribution

In [9]:
# Single attribute
attribute_distribution(df, "size", True)

Frequency  Percentage
Value                       
3         544435   61.210505
22        228839   25.728233
21         97465   10.957932
2          13144    1.477772
1           5564    0.625557

In [10]:
# # All attributes
# for column_name in df.columns:
#     print(attribute_distribution(df, column_name, False))



### (e) Try to describe the data set in just a few sentences. How is the data provided? Which kind of attributes are contained in the data set? How large is the data set in terms of the number of those elements (teams, matches, players, historic data, extra records, and so on)?

The OSHA Injury Tracking Application (ITA) dataset is a structured repository of work-related injury and illness records reported by establishments with 100 or more employees in high-hazard industries. The data is provided as a CSV file and contains attributes related to establishments (e.g., name, location, industry), incidents (e.g., date, type, outcome, days away from work), and employee roles (e.g., job title, Standard Occupation Codes). Additional narrative fields describe incidents and injuries in detail. The dataset size depends on the reporting frequency but typically includes thousands of records, each representing a unique incident, with detailed fields linking establishments, incidents, and employees for comprehensive analysis.

### (f) Analyze the errors and missing values. Write a function to count how many missing values per attribute and per entry you have. Analyze what are the most relevant missing values that might hinder the analysis according to you.

In [11]:
def analyze_and_plot_missing_values(dataframe):
    """
    Analyzes and plots the missing values in the dataset.

    Parameters:
        dataframe (pd.DataFrame): The DataFrame containing the dataset.

    Returns:
        dict: A dictionary with:
              - Total missing values per attribute
              - Percentage of missing values per attribute
              - Missing values per entry
    """
    # Count missing values per attribute
    missing_per_attribute = dataframe.isnull().sum()
    percent_missing_per_attribute = (missing_per_attribute / len(dataframe)) * 100

    # Combine counts and percentages into a DataFrame
    attribute_analysis = (
        pd.DataFrame(
            {
                "Attribute": dataframe.columns,
                "Missing_Count": missing_per_attribute,
                "Percentage_Missing": percent_missing_per_attribute,
            }
        )
        .query("Missing_Count > 0")
        .sort_values(by="Percentage_Missing", ascending=False)
        .set_index("Attribute")
    )  # Sort in descending order

    # Count missing values per entry (row)
    missing_per_entry = dataframe.isnull().sum(axis=1)
    # Distribution of rows by the number of missing values
    row_missing_distribution = missing_per_entry.value_counts().reset_index()
    row_missing_distribution.columns = ["Missing_Count", "Row_Count"]
    row_missing_distribution = row_missing_distribution.sort_values(
        by="Missing_Count", ascending=True
    ).set_index("Missing_Count")

    # Plot missing values per attribute
    fig_attr = px.bar(
        attribute_analysis.reset_index(),
        x="Attribute",
        y="Percentage_Missing",
        text="Percentage_Missing",
        title="Missing Values Per Attribute (Sorted by Percentage)",
        labels={
            "Attribute": "Attribute",
            "Percentage Missing": "Percentage Missing (%)",
        },
    )
    fig_attr.update_traces(texttemplate="%{text:.2f}%", textposition="outside")
    fig_attr.update_layout(
        xaxis=dict(title="Attributes", tickangle=45),
        yaxis=dict(title="Percentage Missing (%)"),
        uniformtext_minsize=8,
        uniformtext_mode="hide",
        showlegend=False,
    )
    fig_attr.show()

    # Plot distribution of missing values per row
    fig_row = px.bar(
        row_missing_distribution.reset_index(),
        x="Missing_Count",
        y="Row_Count",
        text="Row_Count",
        title="Distribution of Missing Values Per Row",
        labels={
            "Missing_Count": "Number of Missing Values",
            "Row_Count": "Number of Rows",
        },
    )
    fig_row.update_traces(texttemplate="%{text}", textposition="outside")
    fig_row.update_layout(
        xaxis=dict(title="Number of Missing Values"),
        yaxis=dict(title="Number of Rows"),
        uniformtext_minsize=8,
        uniformtext_mode="hide",
        showlegend=False,
    )
    fig_row.show()

    return attribute_analysis, row_missing_distribution


# Example usage
df_missing_attributes, df_row_missing_distribution = analyze_and_plot_missing_values(df)

In [12]:
df_missing_attributes

Missing_Count  Percentage_Missing
Attribute                                              
date_of_death                889168           99.968632
time_started_work            111891           12.579839
time_of_incident             110968           12.476067
ein                           76525            8.603660
industry_description          59799            6.723166
company_name                  49056            5.515337
job_description                3654            0.410817
time_unknown                   2049            0.230368
establishment_type             1884            0.211817
naics_year                     1273            0.143123
case_number                      13            0.001462
street_address                    1            0.000112

In [13]:
df_row_missing_distribution.query("Missing_Count > 0")

Row_Count
Missing_Count           
1                 603667
2                 178776
3                  85337
4                  18306
5                   3134
6                     25
7                      1

##### Analysis
1. **`date_of_death` (99.97% missing)**
- **Impact**: This field is crucial for analyzing fatalities but is practically unusable due to the high missing percentage.
- **Recommendation**: Use the `incident_outcome` field, which includes death as a category, to indirectly analyze fatality-related trends.

2. **`time_started_work` (12.58% missing) and `time_of_incident` (12.48% missing)**
- **Impact**: These fields are essential for analyzing temporal trends, such as incidents occurring shortly after starting work. Missing values reduce the reliability of time-dependent analyses.
- **Recommendation**: Focus analyses on the available data or consider imputing missing values based on similar cases or statistical methods.

3. **`ein` (8.60% missing)**
- **Impact**: The EIN uniquely identifies establishments and is vital for merging datasets or conducting establishment-specific studies. Missing values hinder these analyses.
- **Recommendation**: Use `establishment_ID` as an alternative identifier if it is complete.

4. **`industry_description` (6.72% missing)**
- **Impact**: Industry classification is critical for sector-specific risk analysis. Missing values hinder comparisons of workplace safety across industries.
- **Recommendation**: Use `naics_code` for industry-level analysis or group missing values into an "Unknown" category.

5. **`job_description` (0.41% missing)**
- **Impact**: This field is important for analyzing risks associated with specific job roles. Missing data limits occupation-specific insights.
- **Recommendation**: Exclude rows with missing `job_description` from job-specific analyses or impute values based on similar cases.


## Exercise 2 – Goal - Data (Domain specific)


#### General Overall Goal
The primary goal of the visualization tool is to **enable workplace safety analysts, policymakers, and industry leaders** to:
1. Identify trends in workplace injuries and illnesses across industries, establishments, and job roles.
2. Explore the temporal, geographic, and sector-specific distribution of incidents to identify patterns and potential risk factors.
3. Facilitate decision-making by highlighting areas that require safety interventions, such as industries with high incident rates or recurring issues in specific job roles.

#### Target Users
The visualization tool is designed for:
- **Workplace Safety Analysts**: To understand patterns in workplace incidents and investigate contributing factors.
- **Policymakers**: To design and evaluate regulations that mitigate risks in high-hazard industries.
- **Industry Leaders/Managers**: To assess their establishments’ performance compared to industry benchmarks and implement targeted safety measures.

#### Overall Goal and High-Level Actions
The primary goal is **"Exploratory Analysis"**, focusing on:
1. **Comparative Analysis**: Compare incidents across industries, job roles, and geographic regions to identify high-risk categories.
2. **Trend Identification**: Examine temporal trends in incident occurrences and severity (e.g., time of day, seasonality).
3. **Insight Generation**: Drill down into specific establishments or job types to identify recurring patterns or anomalies.
4. **Communication and Awareness**: Present findings in an intuitive, interactive format to raise awareness and drive action.

#### Why This Goal is Suitable for the Available Data
- The dataset contains a wealth of detailed information about workplace incidents, including establishment-level, incident-level, and job-level attributes. These can be visualized to uncover patterns and correlations that would be hard to identify otherwise.
- While some attributes have missing values (e.g., `date_of_death`, `time_started_work`), the remaining data is sufficient to provide meaningful insights at industry, establishment, and incident levels.

#### Why Visualization is the Right Means
1. **Pattern Recognition**: Visualization allows users to recognize patterns and outliers, such as industries with unusually high incident rates.
2. **Exploration and Interaction**: An interactive tool enables users to explore the dataset from various perspectives (e.g., filtering by industry, geographic location, or incident severity).
3. **Decision Support**: Visualizing data enables managers and policymakers to make informed decisions quickly by presenting complex data in an understandable format.
4. **Communication**: Visualizations can convey insights effectively to diverse stakeholders, including non-technical audiences.

This tool is ideal for leveraging the available data to inform workplace safety improvements, reduce incident rates, and ensure compliance with regulations.


## Exercise 3 – Data (What) Domain specific


### (a) Write in section What (Data) the description of the data. You can base it on the analysis you have done in exercise 1. What are the general properties of the data you want to use? 


#### Attributes Needed for the Analysis and Their Relevance

1. **Establishment-Level Attributes**:
   - **`naics_code` and `industry_description`**: Essential for identifying high-risk industries and understanding sector-specific trends.
   - **`state`, `city`, `zip_code`**: Crucial for geographic analysis to detect regional patterns or disparities in workplace safety.
   - **`size`, `annual_average_employees`, and `total_hours_worked`**: Provide context for scaling incident data (e.g., incidents per employee or hours worked) to make comparisons meaningful across establishments of different sizes.

2. **Incident-Level Attributes**:
   - **`date_of_incident`**: Key for temporal analysis, such as identifying trends over time or seasonal variations.
   - **`type_of_incident`**: Important for categorizing and understanding the nature of incidents (e.g., injuries vs. illnesses).
   - **`incident_outcome`**: Crucial for evaluating the severity of incidents and prioritizing interventions.
   - **`dawf_num_away` and `djtr_num_tr`**: Provide metrics to assess the impact of incidents on productivity and employee health.

3. **Narrative and Job-Level Attributes**:
   - **`job_description`**: Helps identify which roles are most vulnerable to workplace incidents, enabling targeted interventions.
   - **`SOC_code` and `SOC_description`**: Provide standard classifications for jobs, supporting cross-industry comparisons.
   - **`incident_description` and `nar_what_happened`**: Offer qualitative insights into incident causes and circumstances, which are valuable for designing preventative measures.

4. **System Metadata**:
   - **`year_of_filing`**: Allows analysis of data trends across multiple reporting years.
   - **`created_timestamp`**: Ensures timeliness and relevance of the data used for analysis.

#### Why These Attributes Are Relevant
These attributes enable comprehensive analyses aligned with the goals of the visualization tool:
- **Comparative Analysis**: Attributes like `naics_code`, `state`, and `incident_outcome` allow comparisons across industries, regions, and severity levels.
- **Trend Identification**: Temporal attributes such as `date_of_incident` and `year_of_filing` help identify trends in workplace safety over time.
- **Actionable Insights**: Narrative and job-level attributes (`job_description`, `SOC_code`, and `incident_description`) provide detailed insights into specific incident causes, enabling targeted interventions.
- **Scalability**: Workforce metrics (`size`, `annual_average_employees`, `total_hours_worked`) ensure that analyses are normalized, allowing for meaningful comparisons across establishments of varying sizes.

By focusing on these attributes, the visualization tool can deliver actionable insights to workplace safety analysts, policymakers, and industry leaders.


### (b) Most of the data sets contain noise, missing data values, and relations, or measurement errors. The data of this course is no exception. In exercise 1, you already looked at the missing values. How will you handle missing data values or measurement errors? Think of multiple ways and their pros and cons.

In [14]:
def preprocess_dataframe(df: pd.DataFrame) -> pd.DataFrame:
    """
    Preprocesses the given DataFrame by performing data cleaning, type conversions,
    and mapping of categorical variables for better interpretability and analysis.

    Parameters:
        df (pd.DataFrame): The input DataFrame containing raw data.

    Returns:
        pd.DataFrame: A cleaned and preprocessed DataFrame with the following transformations:
            - String columns are stripped of whitespace, tabs, and excess spaces.
            - Numeric columns are converted to appropriate numeric types with downcasting.
            - Categorical columns are mapped to meaningful labels and converted to 'category' type.
            - Date and time columns are converted to datetime or time objects as needed.
            - Missing values are imputed based on column-specific logic or filled with default placeholders.
            - Invalid or placeholder values in specific columns (e.g., EIN, soc_code) are replaced with standardized values.
            - Columns with redundant or irrelevant information (e.g., 'year_filing_for') are dropped.

    Key Transformations:
        - String Cleaning: Strips leading/trailing whitespace, removes tabs, and normalizes spacing.
        - Numeric Conversion: Downcasts numeric columns to optimize memory usage.
        - Categorical Mapping: Maps numeric or placeholder codes to meaningful labels.
        - Date/Time Conversion: Parses date and time strings into appropriate formats.
        - Missing Value Imputation:
            - 'case_number', 'company_name', 'street_address', 'job_description': Filled with "Not provided".
            - 'industry_description': Filled with "No description given".
            - 'ein': "Enter EIN" replaced with "No EIN Given".
            - 'time_unknown': Mapped to "No" by default if missing.
            - Others: Column-specific imputation logic applied.
        - Dropped Columns: 'year_filing_for'.

    Notes:
        - Assumes specific formats for date and time columns.
        - Handles invalid or missing data gracefully using pandas' built-in capabilities (e.g., `errors='coerce'`).
    """

    df_copy = df.copy()

    # Define mappings and preprocessing rules
    # TODO: do we care about string size optimization like U-#?
    # TODO: ein - 36-283962 and 74-187392 cause they have a dash instead of a number
    # TODO: street_address - would be nice to extract city, house number
    # TODO: and etc, but format is too different, so idk how to do that
    # TODO: case_number - completely random with numbers and dates and etc.
    # TODO: soc_code - idk how to efficiently store it cause it nn-nnnn format
    # TODO: and 0000 and 9999 for wrong and insufficient info
    # TODO: soc_description are sometimes comma separated for multicategories
    to_string = [
        "establishment_id",
        "ein",
        "company_name",
        "street_address",
        "city",
        "zip_code",
        "industry_description",
        "case_number",
        "job_description",
        "soc_code",
        "soc_description",
        "establishment_name",
        "naics_code",
    ]
    to_numeric = [
        "annual_average_employees",
        "total_hours_worked",
        "dafw_num_away",
        "djtr_num_tr",
    ]
    categorical_mappings = {
        "establishment_type": {
            # TODO: I couldn't find info on 0.0, but it can mean something else
            0.0: "Invalid entry",
            1.0: "Private industry",
            2.0: "State government entity",
            3.0: "Local government entity",
        },
        # TODO: I value interpretability more than pseudo ordering
        # TODO: so I don't encode it as numbers, but explicit categories
        "size": {1: "<20", 2: "20-249", 21: "20-99", 22: "100-249", 3: "250+"},
        "incident_outcome": {
            1: "Death",
            2: "Days away from work (DAFW)",
            3: "Job transfer or restriction",
            4: "Other recordable case",
        },
        "type_of_incident": {
            1: "Injury",
            2: "Skin disorder",
            3: "Respiratory condition",
            4: "Poisoning",
            5: "Hearing Loss",
            6: "All other illness",
        },
        "time_unknown": {0: "No", 1: "Yes"},
        "soc_reviewed": {0: "Not reviewed", 1: "Reviewed", 2: "Not SOC coded"},
        "state": {
            "PA": "Pennsylvania",
            "GA": "Georgia",
            "VA": "Virginia",
            "TX": "Texas",
            "UT": "Utah",
            "AZ": "Arizona",
            "IN": "Indiana",
            "TN": "Tennessee",
            "WI": "Wisconsin",
            "NC": "North Carolina",
            "NY": "New York",
            "OH": "Ohio",
            "IA": "Iowa",
            "AK": "Alaska",
            "OK": "Oklahoma",
            "MN": "Minnesota",
            "MO": "Missouri",
            "IL": "Illinois",
            "CT": "Connecticut",
            "NE": "Nebraska",
            "LA": "Louisiana",
            "WV": "West Virginia",
            "NM": "New Mexico",
            "CO": "Colorado",
            "FL": "Florida",
            "CA": "California",
            "MD": "Maryland",
            "AL": "Alabama",
            "KY": "Kentucky",
            "MI": "Michigan",
            "SC": "South Carolina",
            "ID": "Idaho",
            "KS": "Kansas",
            "MS": "Mississippi",
            "AR": "Arkansas",
            "NV": "Nevada",
            "NH": "New Hampshire",
            "VT": "Vermont",
            "NJ": "New Jersey",
            "DE": "Delaware",
            "MA": "Massachusetts",
            "ND": "North Dakota",
            "WA": "Washington",
            "OR": "Oregon",
            "ME": "Maine",
            "SD": "South Dakota",
            "MT": "Montana",
            "PR": "Puerto Rico",
            "RI": "Rhode Island",
            "WY": "Wyoming",
            "HI": "Hawaii",
            "DC": "District of Columbia",
            "VI": "U.S. Virgin Islands",
            "GU": "Guam",
            "MP": "Northern Mariana Islands",
            "AS": "American Samoa",
        },
    }

    # Convert to string and clean text
    df_copy[to_string] = (
        df_copy[to_string]
        .astype("string")
        .apply(
            lambda col: col.str.strip()
            .str.replace(r"\t", "", regex=True)
            .str.replace(r"\s+", " ", regex=True)
        )
    )

    # Convert to numeric with downcasting
    df_copy[to_numeric] = df_copy[to_numeric].apply(
        pd.to_numeric, errors="coerce", downcast="integer"
    )
    # TODO: soc_probability = 5 means that the soc_code is wrong, which was determined
    # after it was soc_reviewed. What to do with 5?
    df_copy["soc_probability"] = df_copy["soc_probability"].apply(
        pd.to_numeric, errors="coerce", downcast="float"
    )

    # Map categorical columns
    for col, mapping in categorical_mappings.items():
        if col in df_copy:
            df_copy[col] = (
                df_copy[col].map(mapping).fillna("Not stated").astype("category")
            )

    # Handle specific columns
    df_copy["case_number"] = df_copy["case_number"].fillna("Not provided")
    df_copy["company_name"] = df_copy["company_name"].fillna("Not provided")
    df_copy["street_address"] = df_copy["street_address"].fillna("")
    df_copy["naics_year"] = (
        df_copy["naics_year"].fillna("Invalid NAICS codes").astype("category")
    )
    df_copy["ein"] = (
        df_copy["ein"].str.replace("Enter EIN", "No EIN Given").fillna("No EIN Given")
    )
    df_copy["industry_description"] = df_copy["industry_description"].fillna(
        "No description given"
    )
    df_copy["job_description"] = df_copy["job_description"].fillna("No job description")
    df_copy["soc_code"] = (
        df_copy["soc_code"].replace("0000", "00-0000").replace("9999", "99-9999")
    )

    # Date and time conversions
    # TODO: what am I doing with all missing time data?
    date_columns = {
        "date_of_incident": "%m/%d/%Y",
        "date_of_death": "%m/%d/%Y",
        "created_timestamp": "%d%b%y:%H:%M:%S",
    }
    for col, fmt in date_columns.items():
        df_copy[col] = pd.to_datetime(df_copy[col], format=fmt, errors="coerce")

    # TODO: object and not time object. Can be saved with the date though....
    # .dt.time for time object saved as object
    time_columns = ["time_started_work", "time_of_incident"]
    for col in time_columns:
        df_copy[col] = pd.to_datetime(
            df_copy[col], format="%H:%M:%S.%f", errors="coerce"
        )

    # Since it's just 2023, and it does not provide any info
    df_copy = df_copy.drop(columns="year_filing_for")
    # Outlier values are contained by default [172307584.0, 126.0]
    # and [307584.0, 273751.0]. The number of employees is taken from
    # https://www.zippia.com/golden-state-foods-careers-24869/demographics/
    # and then hours are imputed for roughly same scale companies
    df_copy.loc[
        df["company_name"] == "Golden State Foods",
        ["annual_average_employees", "total_hours_worked"],
    ] = [
        4000,
        df_copy.query("3000 < annual_average_employees < 5000")[
            "total_hours_worked"
        ].median(),
    ]

    return df_copy


df_preproc = preprocess_dataframe(df)

In [15]:
df_preproc.sample(5)

establishment_id                                 establishment_name  \
id                                                                           
451654           485945                     Ochsner Baptist Medical Center   
869601          1215585                                            WM 3231   
686265            62625  Newport News Shipbuilding - Division of Huntin...   
162968          1105361                                 INFLD - PLAINFIELD   
800058          1235426                             Mission Inn and Resort   

              ein                             company_name  \
id                                                           
451654  720502505                    OCHSNER HEALTH SYSTEM   
869601  710415188                             Walmart Inc.   
686265  540318880            Huntington Ingalls Industries   
162968  582480149               United Parcel Service Inc.   
800058  843655810  MMI Hotel Group Employment Services LLC   

                street_address                city      state zip_code  \
id                                                                       
451654       2700 Napoleon Ave         New Orleans  Louisiana    70115   
869601          551 LIBERTY DR           GREENWOOD   Arkansas    72936   
686265  4101 Washington Avenue        Newport News   Virginia    23607   
162968     10095 BRADFORD ROAD          PLAINFIELD    Indiana    46168   
800058             10400 FL-48  Howey-In-The-Hills    Florida    34737   

       naics_code naics_year  \
id                             
451654     622110       2022   
869601     452910       2012   
686265     336611       2022   
162968     492110       2022   
800058     721110       2022   

                                     industry_description establishment_type  \
id                                                                             
451654             Hospital, general medical and surgical   Private industry   
869601                   Warehouse Clubs and Supercenters   Private industry   
686265  Shipyard (i.e., facility capable of building s...   Private industry   
162968             Couriers and Express Delivery Services   Private industry   
800058  Hotel management services (i.e., providing man...   Private industry   

           size  annual_average_employees  total_hours_worked case_number  \
id                                                                          
451654     250+                      1448             2295331  AAAC000000   
869601  100-249                       225              391993    C3594995   
686265     250+                     26559            53631727  2023002072   
162968     250+                      1860             2744307  2803269891   
800058  100-249                       234              359159           6   

              job_description soc_code  \
id                                       
451654  LPNHospital Inpatient  99-9999   
869601        STOCKER GROCERY  53-7065   
686265                 RIGGER  49-9096   
162968     LoaderUnloader Hub  53-7062   
800058           Greenskeeper  11-9081   

                                          soc_description   soc_reviewed  \
id                                                                         
451654                                            Uncoded  Not SOC coded   
869601                         Stockers and Order Fillers       Reviewed   
686265                                            Riggers       Reviewed   
162968  Laborers and Freight, Stock, and Material Move...       Reviewed   
800058                                   Lodging Managers   Not reviewed   

        soc_probability date_of_incident             incident_outcome  \
id                                                                      
451654         0.000000       2023-10-27        Other recordable case   
869601         0.999391       2023-04-30        Other recordable case   
686265         0.999949       2023-05-31   Days away from wo

In [16]:
df_preproc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889447 entries, 446827 to 900939
Data columns (total 31 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   establishment_id          889447 non-null  string        
 1   establishment_name        889447 non-null  string        
 2   ein                       889447 non-null  string        
 3   company_name              889447 non-null  string        
 4   street_address            889447 non-null  string        
 5   city                      889447 non-null  string        
 6   state                     889447 non-null  category      
 7   zip_code                  889447 non-null  string        
 8   naics_code                889447 non-null  string        
 9   naics_year                889447 non-null  category      
 10  industry_description      889447 non-null  string        
 11  establishment_type        889447 non-null  category      
 12  si

In [17]:
df_preproc.query("soc_probability != 5 & soc_reviewed == 'Reviewed'")

establishment_id             establishment_name           ein  \
id                                                                     
446827            41940   AristaCare at Meadow Springs     204755042   
448949            41940   AristaCare at Meadow Springs     204755042   
450502            41940   AristaCare at Meadow Springs     204755042   
450647            41940   AristaCare at Meadow Springs     204755042   
123561            41952       SMM Group Chesapeake, VA     453807558   
...                 ...                            ...           ...   
883956          1240511               Huth Ben Pearson     465766195   
883969          1240575  Rolette Community Care Center  No EIN Given   
883972          1240575  Rolette Community Care Center  No EIN Given   
883994          1240575  Rolette Community Care Center  No EIN Given   
883970          1240577   Kincaid Construction Company     330344291   

                            company_name           street_address  \
id                                                                  
446827  AristaCare at Meadow Springs LLC      845 Germantown Pike   
448949  AristaCare at Meadow Springs LLC      845 Germantown Pike   
450502  AristaCare at Meadow Springs LLC      845 Germantown Pike   
450647  AristaCare at Meadow Springs LLC      845 Germantown Pike   
123561             Sims Metal Management        4300 Buell Street   
...                                  ...                      ...   
883956                  Huth Ben Pearson             260 Grant St   
883969     Rolette Community Care Center         804 State Street   
883972     Rolette Community Care Center         804 State Street   
883994     Rolette Community Care Center         804 State Street   
883970      Kincaid Construction Company  4063 Birch Street, #210   

                    city         state zip_code naics_code naics_year  \
id                                                                      
446827  Plymouth Meeting  Pennsylvania    19462     623110       2022   
448949  Plymouth Meeting  Pennsylvania    19462     623110       2022   
450502  Plymouth Meeting  Pennsylvania    19462     623110       2022   
450647  Plymouth Meeting  Pennsylvania    19462     623110       2022   
123561        Chesapeake      Virginia    23324     423930       2022   
...                  ...           ...      ...        ...        ...   
883956          Hartford     Wisconsin    53027     332710       2022   
883969           Rolette  North Dakota    58366     623110       2022   
883972           Rolette  North Dakota    58366     623110       2022   
883994           Rolette  North Dakota    58366     623110       2022   
883970     Newport Beach    California    92660     238130       2022   

                              industry_description establishment_type  \
id                                                                      
446827                  Skilled nursing facilities   Private industry   
448949                  Skilled nursing facilities   Private industry   
450502                  Skilled nursing facilities   Private industry   
450647                  Skilled nursing facilities   Private industry   
123561  Metal scrap and waste merchant wholesalers   Private industry   
...                                            ...                ...   
883956                               Machine shops   Private industry   
883969                  Skilled nursing facilities   Private industry   
883972                  Skilled nursing facilities   Private industry   
883994                  Skilled nursing facilities   Private industry   
883970                         Framing contractors   Private industry   

           size  annual_average_employees  total_hours_worked case_number  \
id                                                                          
446827     250+                       282              357675           1   
448949     250+                       282

In [18]:
output_directory = "datasets"
os.makedirs(output_directory, exist_ok=True)
df.to_parquet(os.path.join(output_directory, "processed_data.parquet"), index=False)

### (c) (Data (What)) Choose one of the methods and implement it for the data set. Describe it in the section and mention what is the effect on the data.

In [ ]:
df_numeric = df_preproc.select_dtypes(include=[np.number])

for column in df_numeric.columns:
    fig_incident = px.histogram(
        df_numeric, x=column, title=f"Histogram of {column}", nbins=50, text_auto=True
    )
    fig_incident.show()

In [ ]:
df_time = df_preproc[df_preproc.select_dtypes(include=["datetime"]).columns]

for column in df_time.columns:
    fig_incident = px.histogram(
        df_time[df_time[column].notna()],
        x=column,
        title=f"Histogram of {column}",
        nbins=100,
    )
    fig_incident.show()

In [ ]:
df_categorical = df_preproc.select_dtypes(include=["category"])

# Generate histograms for categorical data
for column in df_categorical.columns:
    fig = px.histogram(
        df_categorical[column].dropna(),
        x=column,
        title=f"Distribution of {column}",
        text_auto=True,
    )
    fig.show()

# El problemo: wtf is happening to company employees and hours


In [22]:
df_preproc.query("annual_average_employees > 3500000")

establishment_id                   establishment_name        ein  \
id                                                                       
15894           717879             Pointe Coupee Healthcare  371505879   
15945           717879             Pointe Coupee Healthcare  371505879   
9379            987607  Sterling Machine Technologies, Inc.  251792083   
9417            987607  Sterling Machine Technologies, Inc.  251792083   
9493            987607  Sterling Machine Technologies, Inc.  251792083   
9538            987607  Sterling Machine Technologies, Inc.  251792083   

                              company_name         street_address       city  \
id                                                                             
15894                Plantation Management  1820 False River Road  New Roads   
15945                Plantation Management  1820 False River Road  New Roads   
9379   Sterling Machine Technologies, Inc.       220 Kreider Road    Palmyra   
9417   Sterling Machine Technologies, Inc.       220 Kreider Road    Palmyra   
9493   Sterling Machine Technologies, Inc.       220 Kreider Road    Palmyra   
9538   Sterling Machine Technologies, Inc.       220 Kreider Road    Palmyra   

              state zip_code naics_code naics_year industry_description  \
id                                                                        
15894     Louisiana    70760     623110       2022        Nursing homes   
15945     Louisiana    70760     623110       2022        Nursing homes   
9379   Pennsylvania    17078     332710       2022        Machine shops   
9417   Pennsylvania    17078     332710       2022        Machine shops   
9493   Pennsylvania    17078     332710       2022        Machine shops   
9538   Pennsylvania    17078     332710       2022        Machine shops   

      establishment_type   size  annual_average_employees  total_hours_worked  \
id                                                                              
15894   Private industry   250+                  16121261                  89   
15945   Private industry   250+                  16121261                  89   
9379    Private industry  20-99                   3572800               74860   
9417    Private industry  20-99                   3572800               74860   
9493    Private industry  20-99                   3572800               74860   
9538    Private industry  20-99                   3572800               74860   

      case_number job_description soc_code  \
id                                           
15894           1            Cook  35-2012   
15945           2            Cook  35-2012   
9379            4       machinist  51-4041   
9417            1       machinist  51-4041   
9493            2          Welder  51-4121   
9538            3       machinist  51-4041   

                                soc_description soc_reviewed  soc_probability  \
id                                                                              
15894          Cooks, Institution and Cafeteria     Reviewed         0.929956   
15945          Cooks, Institution and Cafeteria     Reviewed         0.929956   
9379                                 Machinists     Reviewed         0.999947   
9417                                 Machinists     Reviewed         0.999947   
9493   Welders, Cutters, Solderers, and Brazers     Reviewed         0.999842   
9538                                 Machinists     Reviewed         0.999947   

      date_of_incident            incident_outcome  dafw_num_away  \
id                                                                  
15894       2023-07-16       Other recordable case              0   
15945       2023-10-10  Days away from work (DAFW)             80   
9379        2023-12-11  Days away from work (DAFW)              1   
9417        2023-01-02       Other recordable case              0   
9493        2023-05-01       Other recordable case              0   
9538        2023-10-3

In [23]:
df_preproc.query("total_hours_worked > 2000000000")

establishment_id                         establishment_name        ein  \
id                                                                              
732211           621240                     Villa Maria Post Acute  833248050   
732336           621240                     Villa Maria Post Acute  833248050   
782869           621240                     Villa Maria Post Acute  833248050   
40311            796230            SpringHill Suites Navarre Beach  863927679   
40329            796230            SpringHill Suites Navarre Beach  863927679   
40367            796230            SpringHill Suites Navarre Beach  863927679   
117704           801935                            Ensign Services  741386053   
24434           1072807  Brazos Manufacturing and Welding Services  844114782   
884236          1240684                              Skyline Homes  800556501   
884885          1240684                              Skyline Homes  800556501   
884886          1240684                              Skyline Homes  800556501   
884889          1240684                              Skyline Homes  800556501   
884893          1240684                              Skyline Homes  800556501   
884895          1240684                              Skyline Homes  800556501   
884897          1240684                              Skyline Homes  800556501   
884898          1240684                              Skyline Homes  800556501   
884899          1240684                              Skyline Homes  800556501   
884901          1240684                              Skyline Homes  800556501   
884903          1240684                              Skyline Homes  800556501   
884986          1240684                              Skyline Homes  800556501   
884987          1240684                              Skyline Homes  800556501   
884988          1240684                              Skyline Homes  800556501   
884989          1240684                              Skyline Homes  800556501   
884991          1240684                              Skyline Homes  800556501   
884992          1240684                              Skyline Homes  800556501   
884993          1240684                              Skyline Homes  800556501   

                           company_name        street_address         city  \
id                                                                           
732211     Santa Maria Healthcare, Inc.  425 Barcellus Avenue  Santa MAria   
732336     Santa Maria Healthcare, Inc.  425 Barcellus Avenue  Santa MAria   
782869     Santa Maria Healthcare, Inc.  425 Barcellus Avenue  Santa MAria   
40311   Concord Hospitality Enterprises   8375 Gulf Boulevard      Navarre   
40329   Concord Hospitality Enterprises   8375 Gulf Boulevard      Navarre   
40367   Concord Hospitality Enterprises   8375 Gulf Boulevard      Navarre   
117704   Sedona Trace Health & Wellness       8324 Cameron Rd       Austin   
24434              Brazos Manufacturing          7173 CR 1205    Rio Vista   
884236                    Skyline Homes       99 Horseshoe Rd        Leola   
884885                    Skyline Homes       99 Horseshoe Rd        Leola   
884886                    Skyline Homes       99 Horseshoe Rd        Leola   
884889                    Skyline Homes       99 Horseshoe Rd        Leola   
884893                    Skyline Homes       99 Horseshoe Rd        Leola   
884895                    Skyline Homes       99 Horseshoe Rd        Leola   
884897                    Skyline Homes       99 Horseshoe Rd        Leola   
884898                    Skyline Homes       99 Horseshoe Rd        Leola   
884899                    Skyline Homes       99 Horseshoe Rd        Leola   
884901                    Skyline Homes       99 Horseshoe Rd        Leola   
884903                    Skyline Homes       99 Horseshoe Rd        Leola   
884986                    Skyline Homes       99 Horseshoe Rd        Leola   
884987                

In [24]:
ratio_df = df_preproc.copy()
ratio_df["hours_employee_ratio"] = (
    ratio_df["total_hours_worked"] / ratio_df["annual_average_employees"]
)
ratio_df.query("hours_employee_ratio < 1 | hours_employee_ratio > 4000").sort_values(
    "hours_employee_ratio"
)

establishment_id         establishment_name           ein  \
id                                                                 
442419           997472      PINNACLE DELIVERY LLC     850584177   
648922          1198634           Cadence Delivery     843067827   
648921          1198634           Cadence Delivery     843067827   
648920          1198634           Cadence Delivery     843067827   
648919          1198634           Cadence Delivery     843067827   
...                 ...                        ...           ...   
406618          1155884  Distribution : Sacramento  No EIN Given   
406619          1155884  Distribution : Sacramento  No EIN Given   
456120          1174303     DESIMONE LOGISTICS LLC     843096586   
456118          1174303     DESIMONE LOGISTICS LLC     843096586   
412279          1156059              Pacer Express     842643431   

                 company_name                         street_address  \
id                                                                     
442419  Pinnacle Delivery LLC                       1025 Phar Lap Pl   
648922   Cadence Delivery LLC                    1911 Bent Creek Way   
648921   Cadence Delivery LLC                    1911 Bent Creek Way   
648920   Cadence Delivery LLC                    1911 Bent Creek Way   
648919   Cadence Delivery LLC                    1911 Bent Creek Way   
...                       ...                                    ...   
406618           Not provided                  1420 W National Drive   
406619           Not provided                  1420 W National Drive   
456120           Not provided  333 GAME DRIVE MUNROE FALLS, OH 44262   
456118           Not provided  333 GAME DRIVE MUNROE FALLS, OH 44262   
412279          Pacer Express                        2325 Calate Rdg   

                city           state zip_code naics_code naics_year  \
id                                                                    
442419          Cary  North Carolina    27519     492210       2022   
648922     Mansfield           Texas    76063     492110       2022   
648921     Mansfield           Texas    76063     492110       2022   
648920     Mansfield           Texas    76063     492110       2022   
648919     Mansfield           Texas    76063     492110       2022   
...              ...             ...      ...        ...        ...   
406618    Sacramento      California    95834     424460       2022   
406619    Sacramento      California    95834     424460       2022   
456120  MUNROE FALLS            Ohio    44262     492110       2022   
456118  MUNROE FALLS            Ohio    44262     492110       2022   
412279   San Antonio           Texas    78253     492110       2022   

                          industry_description establishment_type   size  \
id                                                                         
442419     Local Messengers and Local Delivery   Private industry  20-99   
648922  Couriers and Express Delivery Services   Private industry  20-99   
648921  Couriers and Express Delivery Services   Private industry  20-99   
648920  Couriers and Express Delivery Services   Private industry  20-99   
648919  Couriers and Express Delivery Services   Private industry  20-99   
...                                        ...                ...    ...   
406618                    No description given   Private industry    <20   
406619                    No description given   Private industry    <20   
456120                    No description given   Private industry    <20   
456118                    No description given   Private industry    <20   
412279                        Package Delivery   Private industry    <20   

        annual_average_employees  total_hours_worked case_number  \
id                                                                 
442419                        57                   0           6   
648922                        70                   0           6   


In [25]:
analyze_and_plot_missing_values(df_preproc)

(                   Missing_Count  Percentage_Missing
 Attribute                                           
 date_of_death             889168           99.968632
 time_started_work         111891           12.579839
 time_of_incident          110968           12.476067,
                Row_Count
 Missing_Count           
 0                    227
 1                 717123
 2                 121387
 3                  50710)

## Exercise 4 – Data (What) Abstraction

| id | Variable Name               | Description                                                                                                                                                                        | Data Type         |
|----|-----------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------|
| 1  | establishment_id           | Identifier for the establishment.                                                                                                                                                  | string            |
| 2  | establishment_name         | The name of the establishment reporting data.                                                                                                                                       | string            |
| 3  | ein                        | Employer Identification Number (EIN) is also known as Federal Tax Identification Number. Has 9 digit format. If not given, states “No EIN Given”.                                   | string            |
| 4  | company_name               | The name of the company that owns the establishment.                                                                                                                               | string            |
| 5  | street_address             | The street address of the establishment. If not given, states “Not provided”.                                                                                                      | string            |
| 6  | city                       | The city where the establishment is located.                                                                                                                                       | string            |
| 7  | state                      | Full name of the state or territory where the establishment is located.                                                                                                            | category          |
| 8  | zip_code                   | The full zip code for the establishment. Can be converted to numbers, but stored as a string for interpretability.                                                                 | string            |
| 9  | naics_code                 | The North American Industry Classification System (NAICS) code for the establishment. Data use a 2012, 2017, or 2022 NAICS code.                                                  | string            |
| 10 | naics_year                 | The calendar year reflecting the version of NAICS codes used by the establishment [2012, 2017, or 2022]. Invalid NAICS codes are shown as “Invalid NAICS codes”.                   | category          |
| 11 | industry_description       | The industry description for the establishment.                                                                                                                                    | string            |
| 12 | establishment_type         | Type of establishment: Private industry, State government entity, Local government entity.                                                                                         | category          |
| 13 | size                       | The size of the establishment is employer-reported and based on the maximum number of employees who worked there at any point in the year: <20, 20-249, 20-99, 100-249, 250+.      | category          |
| 14 | annual_average_employees   | The annual average number of employees at the establishment. Note: This field should not be summed across cases in an establishment.                                               | int32             |
| 15 | total_hours_worked         | The total hours worked by all employees at the establishment. Note: This field should not be summed across cases in an establishment.                                              | int64             |
| 16 | case_number                | An employer-assigned case number for each unique case (i.e., injured/ill employee).                                                                                                | string            |
| 17 | job_description            | The job title of the injured/ill employee.                                                                                                                                          | string            |
| 18 | soc_code                   | The 2018 Standard Occupation Code (SOC) assigned by the NIOSH Industry and Occupation Computerized Coding System (NIOCCS) or OSHA.                                                  | string            |
| 19 | soc_description            | Text description of the 2018 SOC Code.                                                                                                                                              | string            |
| 20 | soc_reviewed               | Indicator variable as to whether the SOC code was manually reviewed before posting: Not reviewed, Reviewed, Not SOC coded.                                                         | category          |
| 21 | soc_probability            | The score given by the NIOSH Industry and Occupation Computerized Coding System (NIOCCS) for the expected accuracy of the SOC code. Codes assigned by OSHA are given a score of 5. | float64           |
| 22 | date_of_incident           | The date the incident occurred.                                                                                                                                                    | datetime64[ns]    |
| 23 | incident_outcome           | The most serious outcome that occurred: Death, Days away from work (DAFW), Job transfer or restriction, Other recordable case.                                                     | category          |
| 24 | dafw_num_away              | The number of days away from work the employee required to recover from the incident before returning to work.                                                                      | int16             |
| 25 | djtr_num_tr                | The number of days the employee needed to be transferred or reassigned to another job or placed on restricted duty due to the incident.                                             | int16             |
| 26 | type_of_incident           | The type of incident that occurred: Injury, Skin disorder, Respiratory condition, Poisoning, Hearing Loss, All other illness.                                                      | category          |
| 27 | time_started_work          | The time the affected employee started work prior to the incident.                                                                                                                 | datetime64[ns]    |
| 28 | time_of_incident           | The time the incident occurred.                                                                                                                                                    | datetime64[ns]    |
| 29 | time_unknown               | Was the time of the incident unknown? Yes, No.                                                                                                                                      | category          |
| 30 | date_of_death              | The date the death occurred, if applicable.                                                                                                                                        | datetime64[ns]    |
| 31 | created_timestamp          | Timestamp when the record was created.                                                                                                                                              | datetime64[ns]    |